// ngx-tagify/esm2015/lib/tagify.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class TagifyService {
    constructor() {
        this.tagifyMap = new Map();
    }
    /**
     * Adds a tagify instance, so it is available via service. Used internally.
     */
    add(name, tagify) {
        if (this.tagifyMap.get(name)) {
            console.warn(`There already exists a tagify instance with name ${name}!`);
            return;
        }
        this.tagifyMap.set(name, tagify);
    }
    /**
     * Get tagify instance for full access to tagify API.
     */
    get(name) {
        return this.tagifyMap.get(name);
    }
    /**
     * Removes a tagify instance from service. Used internally.
     */
    remove(name) {
        this.tagifyMap.delete(name);
    }
}
TagifyService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TagifyService_Factory() { return new TagifyService(); }, token: TagifyService, providedIn: "root" });
TagifyService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
TagifyService.ctorParameters = () => [];
